src/GafferOSL/OSLObject.cpp

```diff
@@ -96,17 +104,35 @@ bool initContextCompatibility()
 
 const bool g_contextCompatibilityEnabled = initContextCompatibility();
 
+IECore::InternedString g_shaderPassthroughParameterName( "passthrough" );
+
 } // namespace
 
 OSLObject::OSLObject( const std::string &name )
 	:	SceneElementProcessor( name, IECore::PathMatcher::NoMatch )
 {
 	storeIndexOfNextChild( g_firstPlugIndex );
-	addChild( new ShaderPlug( "shader" ) );
+	addChild( new GafferScene::ShaderPlug( "__shader", Plug::In, Plug::Default & ~Plug::Serialisable ) );
 	addChild( new IntPlug( "interpolation", Plug::In, PrimitiveVariable::Vertex, PrimitiveVariable::Invalid, PrimitiveVariable::FaceVarying ) );
 	addChild( new ScenePlug( "__resampledIn", Plug::In, Plug::Default & ~Plug::Serialisable ) );
 	addChild( new StringPlug( "__resampleNames", Plug::Out ) );
 	addChild( new BoolPlug( "__contextCompatibility", Plug::In, true, Plug::Default & ~Plug::AcceptsInputs ) );
+	addChild( new Plug( "primitiveVariables", Plug::In, Plug::Default & ~Plug::AcceptsInputs ) );
+	addChild( new OSLCode( "__outputInternal" ) );
+
+	ClosurePlugPtr codeOut = new ClosurePlug( "out", Plug::Out );
+	outputCombineOSLCode()->outPlug()->addChild( codeOut );
+
+	OSLShaderPtr closureToShaderConverter = new OSLShader( "__closureToShaderConverter" );
```

Comment:
This would require OSLCode to output a "surface" instead of a "shader".  Perhaps this would be useful, but since this is an internal detail, I might prefer to leave it for later unless you're convinced it would be really easy to add?  How would you specify in the UI that you want an OSLCode that is a "surface"?

Reply:
Whereabouts is the "surface" type required? OSL seems to be moving towards [removing shader types entirely](https://groups.google.com/d/msg/osl-dev/bVBZda-UsbI/L32I71UHBgAJ), so I wonder if we're making life harder than we need to somewhere.